File: clients/php/src/Transport/Transport.php
<?php

declare(strict_types=1);

namespace Sharptown\Client\Transport;

use Sharptown\Client\Http\Response;

/**
 * A pluggable transport. Each implementation speaks one of the Sharptown protocols
 * (REST, JSON-RPC, gRPC) but accepts the same canonical {@link TransformRequest}.
 */
interface Transport
{
    public function transform(TransformRequest $request): Response;
}

File: clients/elixir/lib/sharptown/transport.ex
defmodule Sharptown.Transport do
  @moduledoc """
  Behaviour for a pluggable transport. Each implementation speaks one Sharptown protocol
  (REST, JSON-RPC, gRPC) but accepts the same canonical request.

  The request is a map with: `:base_url`, `:headers`, `:filename`, `:content_type`,
  `:bytes`, `:operations`, `:timeout`.
  """

  alias Sharptown.{Error, Response}

  @type request :: %{
          base_url: String.t(),
          headers: [{String.t(), String.t()}],
          filename: String.t(),
          content_type: String.t(),
          bytes: binary(),
          operations: map(),
          timeout: timeout()
        }

  @callback transform(request(), keyword()) :: {:ok, Response.t()} | {:error, Error.t()}
end
